lifecycleScripts/preinstall.js

```diff
@@ -9,22 +9,46 @@ module.exports = function prepareForBuild() {
 
   return exec("npm -v")
     .then(function(npmVersion) {
-      if (npmVersion.split(".")[0] < 3) {
-        console.log("[nodegit] npm@2 installed, pre-loading required packages");
-        return exec("npm install --ignore-scripts");
-      }
+    if (npmVersion.split(".")[0] < 3) {
+      console.log("[nodegit] npm@2 installed, " + 
+                  "pre-loading required packages");
+      return exec("npm install --ignore-scripts");
+    }
 
-      return Promise.resolve();
+    return Promise.resolve();
     })
     .then(function() {
-      if (buildFlags.isGitRepo) {
-        var submodules = require(local("submodules"));
-        var generate = require(local("../generate"));
-        return submodules()
-          .then(function() {
-            return generate();
-          });
-      }
+    if (buildFlags.isGitRepo) {
+      var submodules = require(local("submodules"));
+      var generate = require(local("../generate"));
+      return submodules()
+        .then(function() {
+          return generate();
+        });
+    }
+    })
+    .catch(function(err) {
+      console.log("npm install failed, change to yarn install");
+      return exec("yarn --version")
```

Comment:
It might be better to move this particular block to immediately after running npm install, so you only have to fall into the submodules step once.